sys/riscv/sbi.c

```diff
@@ -0,0 +1,346 @@
+
+typedef struct sbi_ret {
+  long error;
+  long value;
+} sbi_ret_t;
+
+/* SBI implementation info. */
+static u_long sbi_spec_version;
+static u_long sbi_impl_id;
+static u_long sbi_impl_version;
+
+static bool has_time_extension = false;
+static bool has_ipi_extension = false;
+static bool has_rfnc_extension = false;
+static bool has_srst_extension = false;
+static bool has_console_putchar_extension = false;
+static bool has_console_getchar_extension = false;
+
+/* Hardware implementation info. */
+static u_long mvendorid; /* CPU's JEDEC vendor ID */
+static u_long marchid;
+static u_long mimpid;
+
+static sbi_ret_t sbi_call(u_long ext, u_long func, u_long arg0, u_long arg1,
+                          u_long arg2, u_long arg3, u_long arg4) {
+  register register_t a0 __asm("a0") = (register_t)(arg0);
+  register register_t a1 __asm("a1") = (register_t)(arg1);
+  register register_t a2 __asm("a2") = (register_t)(arg2);
+  register register_t a3 __asm("a3") = (register_t)(arg3);
+  register register_t a4 __asm("a4") = (register_t)(arg4);
+  register register_t a6 __asm("a6") = (register_t)(func);
+  register register_t a7 __asm("a7") = (register_t)(ext);
+
+  __asm __volatile("ecall"
+                   : "+r"(a0), "+r"(a1)
+                   : "r"(a2), "r"(a3), "r"(a4), "r"(a6), "r"(a7)
+                   : "memory");
+
+  return (sbi_ret_t){
+    .error = a0,
+    .value = a1,
+  };
+}
+
+#define SBI_CALL0(e, f) SBI_CALL5(e, f, 0, 0, 0, 0, 0)
+#define SBI_CALL1(e, f, a1) SBI_CALL5(e, f, a1, 0, 0, 0, 0)
+#define SBI_CALL2(e, f, a1, a2) SBI_CALL5(e, f, a1, a2, 0, 0, 0)
+#define SBI_CALL3(e, f, a1, a2, a3) SBI_CALL5(e, f, a1, a2, a3, 0, 0)
+#define SBI_CALL4(e, f, a1, a2, a3, a4) SBI_CALL5(e, f, a1, a2, a3, a4, 0)
+#define SBI_CALL5(e, f, a1, a2, a3, a4, a5) sbi_call(e, f, a1, a2, a3, a4, a5)
+
+/*
+ * Base extension.
+ */
+
+static sbi_ret_t sbi_get_spec_version(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_SPEC_VERSION));
+}
+
+static sbi_ret_t sbi_get_impl_id(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_IMPL_ID));
+}
+
+static sbi_ret_t sbi_get_impl_version(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_IMPL_VERSION));
+}
+
+long sbi_probe_extension(long id) {
+  return (SBI_CALL1(SBI_EXT_ID_BASE, SBI_BASE_PROBE_EXTENSION, id).value);
+}
+
+static sbi_ret_t sbi_get_mvendorid(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_MVENDORID));
+}
+
+static sbi_ret_t sbi_get_marchid(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_MARCHID));
+}
+
+static sbi_ret_t sbi_get_mimpid(void) {
+  return (SBI_CALL0(SBI_EXT_ID_BASE, SBI_BASE_GET_MIMPID));
+}
+
+/*
+ * Timer extension.
+ */
+
+void sbi_set_timer(uint64_t val) {
+  sbi_ret_t ret;
+
+  /* Use the TIME legacy replacement extension, if available. */
+  if (has_time_extension)
+    ret = SBI_CALL1(SBI_EXT_ID_TIME, SBI_TIME_SET_TIMER, val);
+  else
+    ret = SBI_CALL1(SBI_SET_TIMER, 0, val);
+
+  assert(ret.error == SBI_SUCCESS);
```

Comment:
Why we can't propagate error and give opportunity to decide how to interpret that?

Reply:
I agree. Done.